docs/engines.md

```diff
@@ -0,0 +1,209 @@
+to be read off the device and written back into the `amber::Buffer`.
+
+When the script is finished the engine destructor will be executed and must
+cleanup any resources created by the engine. It is the job of the embedder
+to shut down the graphics API.
+
+
+## API Layer
+The engine API is described in `src/engine.h`. The `Engine` base class must be
+implemented by the backend engine.
+
+
+### API Methods
+#### `Engine::Create`
+The engines are all created in `src/engine.cc`. The `Engine::Create` method
+will attempt to create the requested engine and return it if possible. When
+adding a new engine a new block needs to be added to this method. When
+`Engine::Create` is complete, the engine will be created but is _not_
+initialized.
+
+
+#### `Initialize`
+The engine is initialized through the `Initialize` method. The initialize will
+accept engine specific configuration as an `EngineConfig` parameter. This allows
+for clients to set configuration data needed for the engine. The assumption is
+that the engine itself does not initialize the graphics API. The API should
+have been provided and all needed information provided in the `EngineConfig`
+object.
+
+A `Delegate` object is also provided to the engine. The delegate object is used
+when the engine needs to talk back to the embedder. For instance, the delegates
+can tell the engine to `LogGraphicsCalls`. The engine then, should, call the
+`Log` method on the delegate for each internal graphics API method called.
+
+If the executing script specified specific features, instance or device
+extensions they will also be provided. The engine can use these as needed.
+
+
+#### `CreatePipeline`
+The `CreatePipeline` method is responsible for creating an engine specific
+version of the given `amber::Pipeline`. Each command which needs a pipeline
+will have the pipeline provided, so the engine must provide a way to go from
+the amber pipeline to the engine pipeline. There can also be multiple pipeline
+lines of a given type.
+
+All information on the pipeline will be provided including shader and buffer
+information.
+
+
+#### `DoClearColor`
+The `DoClearColor` command provides the colour that is to be used for a given
+pipeline when executing the `DoClear` command.
+
+
+#### `DoClearStencil`
+The `DoClearStencil` command provides the value that is to be used for clearing
+the stencil buffer for a given pipeline when `DoClear` is executed.
+
+
+#### `DoClearDepth`
+The `DoClearDepth` command provides the value that is to be used for clearing
+the depth buffer for a given pipeline when `DoClear` is executed.
+
+
+#### `DoClear`
+The `DoClear` command instructs the engine to clear the various colour and
+depth attachments for the given pipeline.
+
+
+#### `DoDrawRect`
+The `DoDrawRect` instructs the engine to draw the given pipeline in the box
+at (x,y) of size (width,height). The buffers must be read back into the backing
+`amber::Buffer` at the end of this method.
+
+
+### `DoDrawArrays`
```

Comment:
This one is at a higher level of header. (3 rather than 4)

Reply:
done